packages/lit-html/src/lib/lit-html.ts

```diff
@@ -12,9 +12,548 @@
+    let regex = textRegex;
+
+    for (let i = 0; i < strings.length - 1; i++) {
+      const s = strings[i];
+      let attrNameEnd = -1;
+      let lastIndex = 0;
+
+      // The conditions in this loop handle the current parse state, and the
+      // assignments to the `regex` variable are the state transitions.
+      while (lastIndex < s.length) {
+        // Make sure we start searching from where we previously left off
+        regex.lastIndex = lastIndex;
+        const match = regex.exec(s);
+        if (match === null) {
+          if (quote === undefined) {
+            attrNameEnd = -1;
+          }
+          break;
+        }
+        lastIndex = regex.lastIndex;
+        if (regex === textRegex) {
+          // TODO (justinfagnani): there are more ways to open a comment in
+          // HTML, like `<?` or `</{`. Figure out whether / how to handle that.
+          if (match[1] === '!--') {
+            regex = commentRegex;
+          } else {
+            if (rawTextElement.test(match[1])) {
+              regex = new RegExp(`<\/${match[1]}`);
+            } else {
+              regex = tagRegex;
+            }
+          }
+        } else if (regex === commentRegex) {
+          regex = textRegex;
+        } else if (
+          regex === tagRegex ||
+          regex === doubleQuoteAttr ||
+          regex === singleQuoteAttr
+        ) {
+          if (match[1] === '>') {
+            regex = textRegex;
+          } else if (quote && (match[0] === '"' || match[0] === "'")) {
+            // End quoted attribute
+            attrNameEnd = -1;
+            regex = tagRegex;
+          } else {
+            attrNameEnd = regex.lastIndex - match[3].length;
+            quote = match[5];
+            regex =
+              quote === undefined
+                ? tagRegex
+                : quote === '"'
+                ? doubleQuoteAttr
+                : singleQuoteAttr;
+          }
+        } else {
+          // Not one of the five state regexes, so we're at the close of a raw a
+          // text element.
+          regex = tagRegex;
+        }
+      }
+
+      html +=
+        regex === textRegex
+          ? s + nodeMarker
+          : (attrNameEnd === -1
+              ? s
+              : s.slice(0, attrNameEnd) + '$lit$' + s.slice(attrNameEnd)) +
+            marker;
+    }
+    this.__element.innerHTML =
+      html + this.__strings[l] + (type === 2 ? '</svg>' : '');
+
+    if (type === SVG_RESULT) {
+      const content = this.__element.content;
+      const svgElement = content.firstChild!;
+      svgElement.remove();
+      content.append(...svgElement.childNodes);
+    }
+
+    // Walk the template to find binding markers and create TemplateParts
+    while ((node = walker.nextNode()) !== null && bindingIndex < l) {
+      if (node.nodeType === 1) {
+        if ((node as Element).hasAttributes()) {
+          const {attributes} = node as Element;
+          for (let i = 0; i < attributes.length; i++) {
```

Comment:
Done in #1212. There I added an array to remember just the attribute names. Remember more info for other parts was a bit more complicated.

Reply:
The point/idea in #1211 was to remove attribute scanning altogether (aka this loop running a regex  over every attribute of every node) by inserting a "has attribute bindings" comment marker after the `>` of a start tag during string parsing since we know exactly the position of `>` as it's already explicitly matched during the state machine. Seems like the only nasty thing about that would be modifying the string during regex'ing, but seems doable.